### PR TITLE
Add more node debug dump functionality.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add ability to dump node inputs/outputs to files, filter nodes, configure behavior with environment variables.

**Motivation and Context**
I wanted to save an intermediate tensor to a file.

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -615,26 +615,35 @@ The Vitis-AI execution provider is only supported on Linux.
...
 
+To specify that node output data should be dumped (to stdout by default), set this environment variable:
+```
+ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1
+```
 
+To specify that node output data should be dumped to files for nodes with name containing "Foo", set these environment variables:
+```
+ORT_DEBUG_NODE_IO_DUMP_OUTPUT_DATA=1
+ORT_DEBUG_NODE_IO_NODE_NAME_FILTER=Foo
```
Comment (ytaous) [inline BUILD.md:644]: >Foo [](start = 35, length = 3)

Do we want to go by case-sensitive or insensitive?

File: onnxruntime/core/framework/debug_node_inputs_outputs_utils.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  // see DumpFlags
+  // Note: When dumping every node, dumping both input and output may be redundant.
+  // Doing that may be more useful when dumping a subset of all nodes.
+  int dump_flags{DumpFlags::ShapeOnly};
+
+  // filters the nodes that are dumped
+  // currently, dumped nodes are nodes whose name contains this value
+  // if empty, dump all nodes
+  std::string node_name_filter{};
+
```
Comment (skottmckay) [inline onnxruntime/core/framework/debug_node_inputs_outputs_utils.h:54]: Can we also have an op_type filter? And maybe support multiple comma/semi-colon separated values  for the name and op_type?

File: BUILD.md (hunk 1 above)
Comment (edgchen1) [reply BUILD.md]: Going with case-sensitive, exact match for now. We can make it fancier later if needed.

File: onnxruntime/core/framework/debug_node_inputs_outputs_utils.cc [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        auto pattern_begin = delimited_patterns.begin();
+        while (true) {
+          const auto pattern_end = std::find(
+              pattern_begin, delimited_patterns.end(),
+              NodeDumpOptions::FilterOptions::kPatternDelimiter);
+
+          if (std::equal(value.begin(), value.end(), pattern_begin, pattern_end)) return true;
+
+          if (pattern_end == delimited_patterns.end()) break;
+
```
Comment (skottmckay) [inline onnxruntime/core/framework/debug_node_inputs_outputs_utils.cc:35]: Any reason to not use unordered_set for the matching?

**Comment by skottmckay** [comment]
Can there be a simple unit test that's only enabled if DEBUG_NODE_INPUTS_OUTPUTS is defined so it's easy for a developer to test any changes in the future?

File: onnxruntime/core/framework/debug_node_inputs_outputs_utils.cc (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/framework/debug_node_inputs_outputs_utils.cc:35]: I wanted to make the usage of the environment variables and direct setting of NodeDumpOptions similar. Now they both use pattern strings. A possible enhancement is to use regular expressions here.

**Review by skottmckay — APPROVED** [review]

**Pull request merged**